parasoft-findings-teamcity-agent/src/main/java/com/parasoft/findings/teamcity/agent/ParasoftFindingsBuildProcess.java

```diff
@@ -156,63 +160,66 @@ private void doWork() {
+        List<ReportParserDescriptor> descriptors = new ArrayList<ReportParserDescriptor>();
         try {
-            reader = xmlInputFactory.createXMLEventReader(xml);
-            while (reader.hasNext()) {
-                XMLEvent event = reader.nextEvent();
-                switch (event.getEventType()) {
-                case XMLStreamConstants.START_ELEMENT:
-                    String toolName = null;
-                    StartElement startElement = event.asStartElement();
-                    @SuppressWarnings("unchecked")
-                    Iterator<Attribute> attributes = startElement.getAttributes();
-                    while (attributes.hasNext()) {
-                        Attribute attribute = attributes.next();
-                        String name = attribute.getName().getLocalPart();
-                        if ("toolName".equals(name)) {
-                            toolName = attribute.getValue();
-                            break;
-                        }
-                    }
-                    if (toolName != null) {
-                        return "SOAtest".equals(toolName) ?
-                                ReportParserTypes.getDescriptor(ReportParserType.SOATEST.name()) :
-                                ReportParserTypes.getDescriptor(ReportParserType.ANALYZERS.name());
-                    }
-                    return null;
-                default:
-                    break;
-                }
+            Document document = getDocument(from);
+            if(checkIfNodeExists(document, "/ResultsSession/CodingStandards/StdViols/StdViol")) {
+                descriptors.add(ReportParserTypes.getDescriptor(ReportParserType.SA_PMD.name()));
             }
-        } catch (XMLStreamException e) {
+
+            if(checkIfNodeExists(document, "/ResultsSession/CodingStandards/StdViols/DupViol")) {
+                descriptors.add(ReportParserTypes.getDescriptor(ReportParserType.SA_PMD_CPD.name()));
+            }
+
+            if(checkIfNodeExists(document, "/ResultsSession/ExecutedTestsDetails[contains(@type,'FT')]")) {
+                descriptors.add(ReportParserTypes.getDescriptor(ReportParserType.SOATEST.name()));
+            }
+
+            if(checkIfNodeExists(document, "/ResultsSession/ExecutedTestsDetails[contains(@type,'UT')]")) {
+                ReportParserTypes.getDescriptor(ReportParserType.ANALYZERS.name());
+            }
+        } catch (Exception e) {
             reportUnexpectedFormat(from);
             LOG.log(Level.SEVERE, e.getMessage(), e);
-        } finally {
-            if (reader != null) {
-                try {
-                    reader.close();
-                } catch (XMLStreamException e) {
-                    LOG.log(Level.SEVERE, e.getMessage(), e);
-                }
-            }
         }
-        return null;
+        return descriptors;
+    }
+
+    private Document getDocument(File file) throws Exception {
+        if(!file.exists()) {
```

Comment:
If the file doesn't exist, does the original plugin report error in the build?
It looks like it should when failed to parse (in original file line 173). And fall into the catch block (I didn't try this).

Also for Nathan's comment here: [CICD-74](https://jira.parasoft.com/browse/CICD-74), one of the cases he mentioned is that the build should report some error when there's no static analysis report generated. Since we are using one field for all reports, we should report some error and fail the build if: 1) report(s) could not be parsed into anything we care (already handled in line 165 of new file); 2) report(s) file doesn't exist

Maybe we should also test to make sure when using wildcard which matches multiple files, what's the original behavior and whether we handle it correctly.